### PR TITLE
Updated the write method

### PR DESCRIPTION
**Pull request by vruddhi-mehta**
_Hi, thank you for your work. We understand that you want to merge your changes and move on from this issue, but we also want to maintain the safety, readability, and testability of our code. Because of this, we kindly ask that you confirm that you have fulfilled the prerequisites for each category of activity before merging your changes._

### If the Pull Request has a dependency update:

- [ ] I have read the Release Notes for the new version (and intermediate versions if the jump would include more than one). _Don't blindly trust that the dependency will honor SEMVER, always read carefully the release notes_
- [ ] Java 8+ support is maintained
- [ ] The new version has no vulnerabilities
- [ ] A team member reviewed the Pull Request.
- [ ] No backward compatibility is broken.
- [ ] If it is a "provided" dependency, it has been checked that if it is suggested, the version number is the same.

### For ALL the Releases:

- [ ] There is a task in GUS for this change.
- [ ] The corresponding Release Notes have been written and shared with the documentation team.
- [ ] The new code has tests and they have before and after methods to be sure that you are working with a clean environment and that you are leaving the environment clean after they finish.
- [ ] If you are using reflection, create a test to call the methods you are invoking, this way, if there is a change in the API, the test will be able to detect it.
- [ ] If the project has a parent pom and modules, the release pipeline only uploads the modules, so, please, make sure that the parent was deployed in the repositories before considering this release finished.
- [ ] Notify in the Slack Channel that the release is complete, so the Docs team can merge the Documentation and Release Notes PR. The docs team always checks in Anypoint Exchange that the connector is released before merging the release notes.
- [ ] Add the connector build to GUS, and assign that build to the GUS ticket.
- [ ] Create a change case request, based on the Change Case Management doc https://docs.google.com/document/d/1tMJlTTZLaXMLiYwxlMBFY1yJwBmejhc4-IP8HAXgDfY/edit#

**NOTE**: _(applies only for Core-Connectors connectors and modules) The release process ends when you merge the pull request that updates the support branch to the new snapshot, please don't forget to do this._

### Patch Version:

- [ ] The Pull Request has been peer-reviewed.
- [ ] No backward compatibility is broken.
- [ ] Documentation: (Required) Release Notes PR with compatibility table. Share it with the Docs team.

### Minor Version:

- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-1.1.1-new-chiper) to coordinate the release of this version with the documentation team and the architects of our team (at least one week before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been done and has been updated with the changes related to the new functionalities.
- [ ] The pull request has been reviewed by a peer and the team leader.
- [ ] No backward compatibility is broken.
- [ ] Public Documentation: Minor releases nearly always require a revision of the documentation because the release usually includes new features as well as bug fixes. Even if there are no user-facing changes, documentation has to be versioned for every minor release.
  - [ ] (Required) Release Notes PR with compatibility table. Share it with the Docs team.
  - [ ] (Likely required) Reference guide. Run the script and share the file with the Docs team.
  - [ ] (Assess) Often a minor release impacts UI and examples, so the documentation (user guides, examples, troubleshooting guides) should be reviewed and updated accordingly. Notify the docs team if any updates are needed.

### Major Version:

- [ ] Have a document with the specification of the release.
- [ ] Create a slack channel (eg: rl-conninfra-sftp-2.0.0) to coordinate the release of this version with the documentation team and the architects of our team (at least one month before the release). Share the spec doc with the docs team.
- [ ] The documentation fork has been made and updated with the changes related to the new functionalities and possible compatibility problems with the previous version.
- [ ] The systems properties have been removed, detailing in the documentation the changes in the default behavior of the product.
- [ ] The TODO comments of the code have been reviewed.
- [ ] A colleague, the team leader, and the team architect have reviewed the Pull Request.
- [ ] Public Documentation: Major releases break backward compatibility and should never be released without documentation, as this breaks the user experience.
  - [ ] (Required) Upgrade and Migration Guide: Every major release must have a detailed upgrade guide (template) with information about what was changed and what steps the user has to take to ensure the connector works as expected. Share the details with the Docs team so they can update the guide.
  - [ ] (Required) Release Notes PR with compatibility table. Share it with the docs team.
  - [ ] (Required) Reference guide. Run the script and share the file with the Docs team.
  - [ ] User guide updates, new screenshots or examples (if applicable), share the updates with the docs team.

### CODE REVIEW

File: src/main/java/org/mule/extension/sftp/internal/operation/SftpCustomWriteDelegate.java [hunk 1]
```diff
@@ -27,6 +26,6 @@ public SftpCustomWriteDelegate(CustomWriteBufferSize customWriteBufferSize) {
   @Override
   public void write(SftpFileSystemConnection fileSystem, String filePath, InputStream content, FileWriteMode mode, boolean lock,
                     boolean createParentDirectories) {
-    fileSystem.write(filePath, content, mode, lock, createParentDirectories, WriteStrategy.CUSTOM, customWriteBufferSize);
+    fileSystem.write(filePath, content, mode, lock, createParentDirectories, customWriteBufferSize);
```
[the remark this answers is not in the document]
Comment (elcoya) [reply src/main/java/org/mule/extension/sftp/internal/operation/SftpCustomWriteDelegate.java:29]: I would not advice that. FileSystem is the connection in this connector. Making it a class level variable would make it a lot harder to replace if the connection is broken or anything like that.

File: src/main/java/org/mule/extension/sftp/internal/operation/SftpOperations.java [hunk 2]
```diff
@@ -158,7 +160,10 @@ public void write(@Connection SftpFileSystemConnection fileSystem,
                     @Content @Summary("Content to be written into the file") InputStream content,
                     @Optional(defaultValue = "true") boolean createParentDirectories,
                     @Optional(defaultValue = "false") boolean lock, @Optional(
-                        defaultValue = "OVERWRITE") @Summary("How the file is going to be written") @DisplayName("Write Mode") FileWriteMode mode) {
+                        defaultValue = "OVERWRITE") @Summary("How the file is going to be written") @DisplayName("Write Mode") FileWriteMode mode,
+                    @Placement(tab = ADVANCED_TAB) @Optional(defaultValue = "STANDARD") WriteStrategy writeStrategy,
+                    @Placement(tab = ADVANCED_TAB) @Optional(
+                        defaultValue = "BUFFER_SIZE_1KB") CustomWriteBufferSize bufferSizeForWriteStrategy) {
```
Comment (elcoya) [inline src/main/java/org/mule/extension/sftp/internal/operation/SftpOperations.java:166]: Please make the default same as the one in the existing implementation (8K)

**Review by elcoya — COMMENTED** [review]
Left some comments. I suggested other changes offline for the Strategy implementation. Check branch `fix/W-15124410-reimplement-strategy` for more details.

File: src/main/java/org/mule/extension/sftp/internal/connection/FileSystem.java [hunk 3]
```diff
@@ -140,11 +141,46 @@ Result<InputStream, A> read(FileConnectorConfig config, String filePath, boolean
...
+   * and considerations as described in the {@link #read(FileConnectorConfig, String, boolean, Long)} (FileConnectorConfig,
+   * String, boolean)} method
+   * This method is a custom write method call that takes the offset value manually.
+   *
+   * @param filePath                the path of the file to be written
+   * @param content                 the content to be written into the file
+   * @param mode                    a {@link FileWriteMode}
+   * @param lock                    whether or not to lock the file
+   * @param createParentDirectories whether or not to attempt creating any parent directories which don't exists.
+   *    * @param bufferSizeForWriteStrategy  a {@link CustomWriteBufferSize}. Defaults to 1024
```
[the remark this answers is not in the document]
Comment (vruddhi-mehta) [reply src/main/java/org/mule/extension/sftp/internal/connection/FileSystem.java:148]: Done

File: src/main/java/org/mule/extension/sftp/internal/operation/SftpOperations.java (hunk 2 above)
Comment (vruddhi-mehta) [reply src/main/java/org/mule/extension/sftp/internal/operation/SftpOperations.java:166]: Done

File: src/main/java/org/mule/extension/sftp/internal/operation/SftpWriteCommand.java [hunk 4]
```diff
@@ -114,4 +90,21 @@ public void write(String filePath, InputStream content, FileWriteMode mode,
     }
   }
 
+  private URI validateFileExists(String filePath, InputStream content, FileWriteMode mode, boolean createParentDirectory) {
```
[the remark this answers is not in the document]
Comment (vruddhi-mehta) [reply src/main/java/org/mule/extension/sftp/internal/operation/SftpWriteCommand.java]: removed the method as it is not replicated now

File: src/main/java/org/mule/extension/sftp/internal/operation/WriteCommand.java [hunk 5]
```diff
@@ -23,7 +22,7 @@ public interface WriteCommand {
   String IS_A_DIRECTORY_MESSAGE = "Is a directory";
 
   /**
-   * Writes a file under the considerations of {@link FileSystem#write(String, InputStream, FileWriteMode, boolean, boolean, WriteStrategy, CustomWriteBufferSize)}
+   * Writes a file under the considerations of {@link FileSystem#write(String, InputStream, FileWriteMode, boolean, boolean, CustomWriteBufferSize)}
```
[the remark this answers is not in the document]
Comment (vruddhi-mehta) [reply src/main/java/org/mule/extension/sftp/internal/operation/WriteCommand.java]: Resolved

File: pom.xml [hunk 6]
```diff
@@ -14,7 +14,7 @@
     <groupId>org.mule.connectors</groupId>
     <artifactId>mule-sftp-connector</artifactId>
     <packaging>mule-extension</packaging>
-    <version>2.3.0-SNAPSHOT</version>
+    <version>2.3.0-Perf-Test-SNAPSHOT</version>
```
Comment (dipeshkumardutta-sfemu) [inline pom.xml:17]: We need to change this rt?

Comment (vruddhi-mehta) [reply pom.xml]: Made this change for performance testing. Reverted it

**Review by elcoya — COMMENTED** [review]
Left two minor comments. Please fix.

File: src/main/java/org/mule/extension/sftp/internal/connection/write/SftpStandardWriter.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    this.sftpClient = sftpClient;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void write(String path, InputStream stream, FileWriteMode mode, URI uri) throws IOException {
+    try (OutputStream out = sftpClient.getOutputStream(path, mode)) {
+      byte[] buf = new byte[8192];
```
[the remark this answers is not in the document]
Comment (vruddhi-mehta) [reply src/main/java/org/mule/extension/sftp/internal/connection/write/SftpStandardWriter.java:36]: Done

**Review by elcoya — DISMISSED** [review]
Looks good

**Review by elcoya — APPROVED** [review]

**Review by dipeshkumardutta-sfemu — APPROVED** [review]
LGTM

**Pull request merged**